core/templates/unordered_hash_map.h

```diff
@@ -0,0 +1,659 @@
+			elements = p_elements;
+			hashes = p_hashes;
+			capacity = p_capacity;
+			index = p_index;
+			if (p_capacity != 0) {
+				// Ensure it sits on a valid hash
+				while (hashes[index] == EMPTY_HASH) {
+					index++;
+					if (unlikely((uint32_t)index == capacity)) {
+						elements = nullptr;
+						hashes = nullptr;
+						capacity = 0;
+						index = -1;
+						return;
+					}
+				}
+			}
+		}
+		_FORCE_INLINE_ operator ConstIterator() {
+			return ConstIterator(elements, hashes, capacity, index);
+		}
+		_FORCE_INLINE_ Iterator() {}
+		_FORCE_INLINE_ Iterator(const Iterator &p_it) {
+			hashes = p_it.hashes;
+			elements = p_it.elements;
+			capacity = p_it.capacity;
+			index = p_it.index;
+		}
+		_FORCE_INLINE_ void operator=(const Iterator &p_it) {
+			hashes = p_it.hashes;
+			elements = p_it.elements;
+			capacity = p_it.capacity;
+			index = p_it.index;
+		}
+
+	private:
+		KeyValue<TKey, TValue> *elements = nullptr;
+		const uint32_t *hashes = nullptr;
+		uint32_t capacity = 0;
+		int32_t index = -1;
+	};
+
+	_FORCE_INLINE_ Iterator begin() {
+		return num_elements ? Iterator(elements, hashes, get_capacity(), 0) : Iterator();
+	}
+	_FORCE_INLINE_ Iterator end() {
+		return Iterator();
+	}
+	_FORCE_INLINE_ Iterator last() {
+		if (num_elements == 0) {
+			return Iterator();
+		}
+		uint32_t index = get_capacity() - 1;
+		while (hashes[index] != EMPTY_HASH) {
+			index--;
+		}
+		return Iterator(elements, hashes, get_capacity(), index);
+	}
+
+	_FORCE_INLINE_ Iterator find(const TKey &p_key) {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return end();
+		}
+		return Iterator(elements, hashes, get_capacity(), pos);
+	}
+
+	_FORCE_INLINE_ void remove(const Iterator &p_iter) {
+		if (p_iter) {
+			erase(p_iter->key);
+		}
+	}
+
+	_FORCE_INLINE_ ConstIterator begin() const {
+		return num_elements ? ConstIterator(elements, hashes, get_capacity(), 0) : ConstIterator();
+	}
+	_FORCE_INLINE_ ConstIterator end() const {
+		return ConstIterator();
+	}
+	_FORCE_INLINE_ ConstIterator last() const {
+		if (num_elements == 0) {
+			return ConstIterator();
+		}
+		uint32_t index = get_capacity() - 1;
+		while (hashes[index] != EMPTY_HASH) {
+			index--;
+		}
+		return ConstIterator(elements, hashes, get_capacity(), index);
+	}
+
+	_FORCE_INLINE_ ConstIterator find(const TKey &p_key) const {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return end();
+		}
+		return Iterator(elements, hashes, get_capacity(), pos);
+	}
+
+	/* Indexing */
+
+	const TValue &operator[](const TKey &p_key) const {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		CRASH_COND(!exists);
+		return elements[pos].value;
+	}
+
+	TValue &operator[](const TKey &p_key) {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return elements[_insert(p_key, TValue())].value;
+		} else {
+			return elements[pos].value;
+		}
```

Comment:
```suggestion
		if (!exists) {
			pos = _insert(p_key, TValue());
		}
		return elements[pos].value;
```

This seems to fix crash when using `map[key] = val` for insertion.